src/common/jq_repository_parser.go

```diff
@@ -0,0 +1,82 @@
+package common
+
+import (
+	"encoding/json"
+	"github.com/opslevel/opslevel-go/v2023"
+	"github.com/rs/zerolog/log"
+	"strings"
+)
+
+type RepositoryDTO struct {
+	Name      string
+	Directory string
+	Repo      string
+}
+
+func (r *RepositoryDTO) Convert() opslevel.ServiceRepositoryCreateInput {
+	return opslevel.ServiceRepositoryCreateInput{
+		Repository:    *opslevel.NewIdentifier(r.Repo),
+		BaseDirectory: r.Directory,
+		DisplayName:   r.Name,
+	}
+}
+
+type JQRepositoryParser struct {
+	programs []*JQFieldParser
+}
+
+func NewJQRepositoryParser(expressions []string) *JQRepositoryParser {
+	programs := make([]*JQFieldParser, len(expressions))
+	for i, expression := range expressions {
+		programs[i] = NewJQFieldParser(expression)
+	}
+	return &JQRepositoryParser{
+		programs: programs,
+	}
+}
+
+func (p *JQRepositoryParser) Run(data string) ([]opslevel.ServiceRepositoryCreateInput, error) {
+	output := make([]opslevel.ServiceRepositoryCreateInput, 0, len(p.programs))
+	for _, program := range p.programs {
+		response, err := program.Run(data)
+		//log.Warn().Msgf("expression: %s\nresponse: %s", program.program.Program, response)
+		if err != nil {
+			log.Warn().Msgf("unable to parse alias from expression: %s", program.program.Program)
+			return nil, err
+		}
+		if response == "" {
+			continue
+		}
+		if strings.HasPrefix(response, "[") && strings.HasSuffix(response, "]") {
+			if response == "[]" {
+				continue
+			}
+			var repos []RepositoryDTO
+			if err := json.Unmarshal([]byte(response), &repos); err == nil {
+				for _, repo := range repos {
+					output = append(output, repo.Convert())
+				}
+			} else {
+				// Try as []string
+				var repoNames []string
+				if err := json.Unmarshal([]byte(response), &repoNames); err == nil {
+					for _, repoName := range repoNames {
+						output = append(output, opslevel.ServiceRepositoryCreateInput{Repository: *opslevel.NewIdentifier(repoName)})
+					}
+				}
+			}
+		} else if strings.HasPrefix(response, "{") && strings.HasSuffix(response, "}") {
+			var repo RepositoryDTO
+			if err := json.Unmarshal([]byte(response), &repo); err != nil {
+				// TODO: log error
+				log.Warn().Err(err).Msgf("unable to marshal repo expression: %s\n%s", program.program.Program, response)
```

Comment:
Can we use the `log` to complete all these `// TODO: log error` comments within this PR? I noticed this TODO has been completed but the rest have not.

Reply:
ya - my error handling is WIP.  I don't remember exactly how the thing calling this works so i need to figure out that refactor first to see how error handling in this needs to work.  Does it need to be fault tolerant and just log errors or does it need to halt the whole process and bubble up.